codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/integration/Waiters.java

```diff
@@ -0,0 +1,835 @@
+                    writer.write("options.MaxAttempts = $L ", DEFAULT_MAX_ATTEMPTS);
+                    writer.write("options.WaiterStateMutator = $L", generateWaiterStateMutatorName(waiterName));
+                    writer.write("");
+
+                    writer.openBlock("for _, fn := range optFns {",
+                            "}", () -> {
+                                writer.write("fn(&options)");
+                            });
+
+                    writer.openBlock("return &$T {", "}", clientSymbol, () -> {
+                        writer.write("client: client, ");
+                        writer.write("options: options, ");
+                    });
+                });
+    }
+
+    /**
+     * Generates waiter invoker functions to call specific operation waiters
+     * These waiter invoker functions is defined on each modeled waiter client.
+     * The invoker function takes in a context, along with operation input, and
+     * optional functional options for the waiter.
+     */
+    private void generateWaiterInvoker(
+            Model model,
+            SymbolProvider symbolProvider,
+            GoWriter writer,
+            OperationShape operationShape,
+            String waiterName,
+            Waiter waiter
+    ) {
+        StructureShape inputShape = model.expectShape(
+                operationShape.getInput().get(), StructureShape.class
+        );
+
+        Symbol operationSymbol = symbolProvider.toSymbol(operationShape);
+        Symbol inputSymbol = symbolProvider.toSymbol(inputShape);
+
+        Symbol waiterOptionsSymbol = SymbolUtils.createPointableSymbolBuilder(
+                generateWaiterOptionsName(waiterName)
+        ).build();
+
+        Symbol clientSymbol = SymbolUtils.createPointableSymbolBuilder(
+                generateWaiterClientName(waiterName)
+        ).build();
+
+        writer.write("");
+        writer.addUseImports(SmithyGoDependency.CONTEXT);
+        writer.writeDocs(
+                String.format("%s calls the waiter function for %s waiter", WAITER_INVOKER_FUNCTION_NAME, waiterName)
+        );
+        writer.openBlock("func (w $P) $L(ctx context.Context, params $P, optFns ...func($P)) error {", "}",
+                clientSymbol, WAITER_INVOKER_FUNCTION_NAME, inputSymbol, waiterOptionsSymbol,
+                () -> {
+                    writer.write("options := $T{}", waiterOptionsSymbol);
+                    writer.write("options.MinDelay = w.options.MinDelay");
+                    writer.write("options.MaxDelay = w.options.MaxDelay");
+                    writer.write("options.MaxWaitTime = w.options.MaxWaitTime");
+                    writer.write("options.MaxAttempts = w.options.MaxAttempts");
+                    writer.write("options.EnableLogger = w.options.EnableLogger");
+                    writer.write("options.WaiterStateMutator = w.options.WaiterStateMutator");
+
```

Comment:
This isn't really needed since `options` is a value, not pointer member. Creating a local variable of `options` from the `w.options` is equivalent to this since all values of the options structure are value.

```go
options := w.options
```

If we end up ever adding pointer values that need deeper copies similar to the API Client's options struct a `clone` method created when needed.`